feat(usage): at-least-once and dynamic sampling (#26)
charlypoly/graphql-ruby-hive: README.md

  use(
    GraphQL::Hive,
    {
      # mandatory
      token: 'YOUR-TOKEN',

      # optional
      enabled: true, # enable/disable Hive Client
      debug: false, # verbose logs
      logger: MyLogger.new,
      endpoint: 'app.graphql-hive.com',
      port: 80, 
      buffer_size: 50, # forward the operations data to Hive every 50 requests

      collect_usage: true, # report usage to Hive
      collect_usage_sampling: {
        # optional members of `collect_usage_sampling`  
        sample_rate: 0.5, # % of operations reported
        sampler: proc { |context| context.operation_name.includes?('someQuery') 1 : 0.5 }, # assign custom sampling rates (overrides `sampling rate`)
        at_least_once: true, # sample every distinct operation at least once
        key_generator: proc { |context| context.operation_name } # assign custom keys to distinguish between distinct operations

>       }

Reviewer: This looks much better!

  use(
    GraphQL::Hive,
    {
      # mandatory
      token: 'YOUR-TOKEN',

      # optional
      enabled: true, # enable/disable Hive Client
      collect_usage: true, # report usage to Hive
      collect_usage_sampling_rate: 1.0, # % of operations reported
      debug: false, # verbose logs
      logger: MyLogger.new,
      endpoint: 'app.graphql-hive.com',
      port: 80, 
      buffer_size: 50, # forward the operations data to Hive every 50 requests

      # to sample every distinct operation at least once, set `at_least_once_sampling_keygen` to `true`
      # alternatively, pass an optional proc to use custom keys to distinguish between distinct operations

>       at_least_once_sampling_keygen: proc { |context| context.operation_name }

Reviewer: I just got a little confused about this property, because it can be `true` or a proc, which are two different types (boolean vs fn). It is allowed in ruby, but I just find it a bit confusing myself 😄 

I wonder if there's a way to improve this. Have we considered an object config instead, like there is one for `reporting`?

For example:
```ruby
sampling: {
  at_least_once_sampling: boolean,
  at_least_once_sampling_keygen: proc { },
  collect_usage_sampling_rate: number,
  collect_usage_sampler: proc { }
}
```

That way, we can have a boolean `true` to just say that we want at least once sampling, and a optional proc if they want to define how to sample it with a keygen.

Author: haha cassidy mentioned the same point while pairing! i've updated the config to an object now, which as you said, makes it much clearer in the code. i do like the idea of combining sampling all into one object 👀👀

    GraphQL::Hive,
    {
      token: 'YOUR-TOKEN',
      collect_usage: true, # report usage to Hive (optional)
      report_schema: true,  # publish schema to Hive (optional)
      enabled: true, # enable/disable Hive Client (optional)
      debug: false, # verbose logs
      logger: MyLogger.new,  # optional
      endpoint: 'app.graphql-hive.com',  # optional
      port: 80,  # optional
      buffer_size: 50, # forward the operations data to Hive every 50 requests (optional)
      collect_usage_sampling: 1.0, # report % of operations (optional)
      reporting: {  # mandatory if `report_schema: true`
        author: 'Author of the latest change', # mandatory member of `reporting`
        commit: 'git sha or any identifier',  # mandatory member of `reporting`
        service_name: '', # optional
        service_url: '', # optional
      },
      # you can pass an optional proc that will help identify the client (ex: Apollo web app) that performed the query
      client_info: Proc.new { |context| { name: context.client_name, version: context.client_version } }

>       # NOTE: the following field overrides collect_usage_sampling
>       # if you want all operations to be sampled at least once, you can pass an optional proc to conditionally assign sampling rates, based on the sampling context of the operation

Reviewer 2: ```suggestion
      # NOTE: the following field overrides collect_usage_sampling
      # if you want all operations to be sampled at least once, you can pass a proc to collect_usage_sampler to conditionally assign sampling rates, based on the context of the operation
```

Is my understanding in this edit correct?

Author: yes! these comments will definitely need to be workshopped on my end

>       collect_usage_sampler: Proc.new { |sampling_context| sampling_context.operation_name.includes?('someQuery') 1 : 0.2 }

Reviewer 2: Nice! This will let us exclude operations from load tests.
